src/covers/discogscoverprovider.cpp

```diff
@@ -0,0 +1,373 @@
+
+  NewClosure(reply, SIGNAL(error(QNetworkReply::NetworkError)), this,
+             SLOT(SearchRequestError(QNetworkReply::NetworkError,
+                                     QNetworkReply*, int)),
+             reply, s_ctx->id);
+  NewClosure(reply, SIGNAL(finished()), this,
+             SLOT(HandleSearchReply(QNetworkReply*, int)), reply, s_ctx->id);
+}
+
+void DiscogsCoverProvider::SendReleaseRequest(
+    DiscogsCoverSearchContext* s_ctx, DiscogsCoverReleaseContext* r_ctx) {
+  typedef QPair<QString, QString> Arg;
+  typedef QList<Arg> ArgList;
+
+  typedef QPair<QByteArray, QByteArray> EncodedArg;
+  typedef QList<EncodedArg> EncodedArgList;
+
+  QUrlQuery url_query;
+  QStringList query_items;
+
+  ArgList args =
+      ArgList() << Arg("key", QByteArray::fromBase64(kAccessKeyB64))
+                << Arg("secret", QByteArray::fromBase64(kSecretKeyB64));
+  // Encode the arguments
+  for (const Arg& arg : args) {
+    EncodedArg encoded_arg(QUrl::toPercentEncoding(arg.first),
+                           QUrl::toPercentEncoding(arg.second));
+    query_items << QString(encoded_arg.first + "=" + encoded_arg.second);
+    url_query.addQueryItem(encoded_arg.first, encoded_arg.second);
+  }
+
+  QUrl url(r_ctx->resource_url);
+
+  // Sign the request
+  const QByteArray data_to_sign =
+      QString("GET\n%1\n%2\n%3")
+          .arg(url.host(), url.path(), query_items.join("&"))
+          .toLatin1();
+  const QByteArray signature(Utilities::HmacSha256(
+      QByteArray::fromBase64(kSecretKeyB64), data_to_sign));
+
+  // Add the signature to the request
+  url_query.addQueryItem("Signature",
+                         QUrl::toPercentEncoding(signature.toBase64()));
+
+  url.setQuery(url_query);
+  QNetworkReply* reply = network_->get(QNetworkRequest(url));
+
+  NewClosure(reply, SIGNAL(error(QNetworkReply::NetworkError)), this,
+             SLOT(ReleaseRequestError(QNetworkReply::NetworkError,
+                                      QNetworkReply*, int, int)),
+             reply, s_ctx->id, r_ctx->id);
+  NewClosure(reply, SIGNAL(finished()), this,
+             SLOT(HandleReleaseReply(QNetworkReply*, int, int)), reply,
+             s_ctx->id, r_ctx->id);
+}
+
+void DiscogsCoverProvider::HandleSearchReply(QNetworkReply* reply, int s_id) {
+  reply->deleteLater();
+
+  DiscogsCoverSearchContext* s_ctx;
+  if (!requests_search_.contains(s_id)) {
+    return;
+  }
+  s_ctx = requests_search_.value(s_id);
+  if (s_ctx == nullptr) return;
+
+  QString json_string = reply->readAll();
+  QByteArray json_bytes = json_string.toLocal8Bit();
+  auto json_doc = QJsonDocument::fromJson(json_bytes);
+  if (json_doc.isNull()) {
+    qLog(Error) << "Discogs: Failed to create JSON doc.";
+    EndSearch(s_ctx);
+    return;
+  }
+  if (!json_doc.isObject()) {
+    qLog(Error) << "Discogs: JSON is not an object.";
+    EndSearch(s_ctx);
+    return;
+  }
+
+  QJsonObject json_obj = json_doc.object();
+  if (json_obj.isEmpty()) {
```

Comment:
Do all of these checks in one conditional

Reply:
Agree